### PR TITLE
feat/message system

### LINKED ISSUE #2752: Remote secure messaging system
Let's design and implement secure remote messaging system

We might need to send various messages to the desktop app, such as:
* emergencies (upcoming hard forks, splits, coin removal etc.)
* desktop app kill switch

Requirements:
* possibility to address different versions differently (version comparison)
* security - messages are signed 

**Product specification:** [Notion](https://www.notion.so/satoshilabs/Secure-messaging-system-8ca41332ce504a0f8f6d74fc4bb9c8ab) 

Modal: [Zeplin](https://app.zeplin.io/project/5ee87c00f6af719a645d14c3/screen/6047b1514e30a0b8c3781a88) 
Context: [Zeplin](https://app.zeplin.io/project/5ee87c00f6af719a645d14c3/screen/6047b15bbafdbb06f6d25f36) 
Banner: [Zeplin](https://app.zeplin.io/project/5ee87c00f6af719a645d14c3/screen/6047b148d44ac2bb6f429603)

### PR DESCRIPTION
**Pull request by tomasklim**
[To be done before freeze - 4 May 2021](https://github.com/trezor/trezor-suite/issues/3694)

[Follow-up issue](https://github.com/trezor/trezor-suite/issues/3693)

### CODE REVIEW

File: packages/suite/src/utils/suite/messageSystem.ts
Comment (szymonlesisz) [inline packages/suite/src/utils/suite/messageSystem.ts]: bowser discussed ip.

File: packages/suite/src/utils/suite/messageSystem.ts
Comment (szymonlesisz) [inline packages/suite/src/utils/suite/messageSystem.ts]: new dependency (semver) concern same as above.
i think i already saw version validation in-house scripts somewhere (@keraf wrote some as i remember correctly) could we recycle them instead?

Comment (szymonlesisz) [reply packages/suite/src/utils/suite/messageSystem.ts]: lol: `yarn why semver` :) so we already have every possible version :)
what is others optinion?

File: packages/suite/src/middlewares/suite/messageSystemMiddleware.ts
Comment (szymonlesisz) [inline packages/suite/src/middlewares/suite/messageSystemMiddleware.ts]: nitpick: you can put all this constant into array:
`const actions = [STORAGE.LOADED, SUITE.SELECT_DEVICE, ...];`
and check via:
`actions.indexOf(action.type)` or `actions.includes(action.type)`

similar to this:
https://github.com/trezor/trezor-suite/blob/develop/packages/suite/src/actions/wallet/selectedAccountActions.ts#L195-L209

File: packages/suite/src/middlewares/suite/messageSystemMiddleware.ts
Comment (szymonlesisz) [inline packages/suite/src/middlewares/suite/messageSystemMiddleware.ts]:
```suggestion
            messages.forEach(message => {
```
nitpick:
message is already typed in messages, no? 1 import less

File: packages/suite/src/middlewares/suite/suiteMiddleware.ts
Comment (szymonlesisz) [inline packages/suite/src/middlewares/suite/suiteMiddleware.ts]: this may be moved to a promise in line 62, fetch witch some short timeout.
potential use case: new config may be downloaded before connect initialization and prevent it

File: packages/suite/src/support/tests/setupJest.ts
Comment (szymonlesisz) [inline packages/suite/src/support/tests/setupJest.ts:347]: this could be maintained in the actual tests where its used. or is there other reason to be shareable?

File: packages/suite/src/actions/suite/constants/messageSystemConstants.ts
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/constants/messageSystemConstants.ts]: tracking TODO

File: packages/suite/src/utils/suite/messageSystem.ts
Comment (tomasklim) [reply packages/suite/src/utils/suite/messageSystem.ts]: These in-house scripts are available in connect library https://github.com/trezor/connect/blob/develop/src/js/utils/versionUtils.js 

It is ready for comparison of two versions. However, in this case, I need to work with ranges, carets, tildes,...

So I believe working library which is used everywhere is the best option

Comment (tomasklim) [reply packages/suite/src/utils/suite/messageSystem.ts]: The in-house version validation is available in connect library https://github.com/trezor/connect/blob/develop/src/js/utils/versionUtils.js

However, it is not sufficient for this use case because it only allows comparing two versions. I need to compare ranges, work with tildes, carets,... 

I looked for alternatives but I did not find any more lightweight.

File: packages/suite/src/utils/suite/messageSystem.ts
Comment (tomasklim) [reply packages/suite/src/utils/suite/messageSystem.ts]: I will replace bowser with ua-parser.js which is used in browser-detection to have it united. 

The solution we talked about is not applicable for the desktop app as browser detection is not available there.

Having an in-house detection module requires a lot of regexes for parsing user agent to correctly detect os and browser versions. This would add a lot of painful code, difficult to maintain, to our project.

Plus, I will create helper functions in env.ts file as we discussed

Comment (tomasklim) [reply packages/suite/src/utils/suite/messageSystem.ts:146]: + I will create helper functions in env.ts file as we discussed

File: packages/suite/src/support/tests/setupJest.ts
Comment (tomasklim) [reply packages/suite/src/support/tests/setupJest.ts:347]: it is used in messageSystemActions test and in fixtures of messageSystem utils test

**Comment by tomasklim** [comment]
@szymonlesisz @keraf @slowbackspace ready to be reviewed again

@tsusanka if you want to check

@prusnak I need you to check [feat(ci): validate, sign config jobs](https://github.com/trezor/trezor-suite/pull/3565/commits/1a4974040a2db090a2be7f1a0cee2d049891506b) and solve TODOs (stick). It is required to write a script to upload signed configs to s3 (data.trezor.io) and generate private/public production keys, save them to key vault and link variables to the ci job. Thank you

File: packages/suite/src/actions/suite/messageSystemActions.ts
Comment (goodhoko) [inline packages/suite/src/actions/suite/messageSystemActions.ts:108]: It would be cool to report to sentry here. It should raise some eyebrows when this fails especially on the authentication or sequence checks. .)

File: packages/suite/src/actions/suite/constants/messageSystemConstants.ts
Comment (goodhoko) [inline packages/suite/src/actions/suite/constants/messageSystemConstants.ts]: Any reason why we don't just
1) fetch the config on app start
2) `setInterval` that would fetch the config every six hours

?

File: packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx
Comment (goodhoko) [inline packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx]: How does this work on desktop? Does it opens a browser?

File: docs/misc/message_system.md
Comment (goodhoko) [inline docs/misc/message_system.md:58]: Does this validation include a check that the generated types in `messageSystem.ts` (as comitted) correspond to the JSON schema? Since we decided to track a generated file in git we should at least ensure it won't lag behind its source.

File: ci/packages/suite-web.yml
Comment (goodhoko) [inline ci/packages/suite-web.yml]: Can't we set the `JWS_PUBLIC_KEY` env var directly without this reassigment?

**Review by goodhoko — APPROVED** [review]
Very nice PR! 🖤 Reads like a book.

Just a few

File: packages/suite/src/actions/suite/constants/messageSystemConstants.ts
Comment (goodhoko) [reply packages/suite/src/actions/suite/constants/messageSystemConstants.ts]: Aaah now I see that we need to shorten the interval when the fetch fails. That makes sense.

I still find these two interval a bit more complicated. What about doing somethings like:

```js
function fetchConfig() {
   try {
     const res = fetchConfigFromServer()
     setTimeout(fetchConfig, 6 hours)
     return res
   } catch {
     setTimeout(fetchConfig, 10 minutes)
}
```

Feel free to keep it as is. RBY. (Resolve by yourself)

**Comment by prusnak** [comment]
I reviewed the parts I was being asked to provide feedback on and sent two commits:
* 7ee8262 - replaced upload with correct `aws s3 cp` commands
* 55f6ce9 - removed JWS_{PUBLIC,PRIVATE}_KEY shenanigans from CI yml files - it should be enough to jest set these two variables in the CI (for codesign branch) and the rest should work automatically

File: ci/packages/suite-web.yml
Comment (prusnak) [reply ci/packages/suite-web.yml]: Right, addressed in 55f6ce9

File: packages/suite/src/actions/suite/constants/messageSystemConstants.ts
Comment (tomasklim) [reply packages/suite/src/actions/suite/constants/messageSystemConstants.ts]: The main reason for this, is possibility of inconsistent behavior of different browsers and operating system during the time when you make your computer sleep and wake up. Another can also be "pausing" tabs in the browser.

Using this 10 minutes interval it is guaranteed, that if you launch app, sleep your computer and wake it up after 6 hours, you will get new config in max 10 minutes.

File: docs/misc/message_system.md
Comment (tomasklim) [reply docs/misc/message_system.md:58]: I wondered if it is good idea to commit types into repo as they are generated by `build:libs` script. This is probably a reason, why it would be better not commit it.

File: packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx
Comment (tomasklim) [reply packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx]: very good point, yes it opens default browser

File: docs/misc/message_system.md
Comment (tsusanka) [inline docs/misc/message_system.md]: I am wondering if this is not too aggressive. Imagine we fuck up, remove the config from our server and then thousands of Suites are talking to our server each 10 minutes. It shouldn't be a problem but to be on the safe side maybe one hour is enough?

File: packages/suite-data/src/message-system/config/config.v1.json
Comment (tsusanka) [inline packages/suite-data/src/message-system/config/config.v1.json:6]: Before merging we need to clean this config up right? I mean it would be used otherwise, correct?

**Review by tsusanka — APPROVED** [review]
I went through CI and readme and it looks great :tada:. I have just two questions.

File: packages/suite/src/actions/suite/constants/messageSystemConstants.ts
Comment (goodhoko) [reply packages/suite/src/actions/suite/constants/messageSystemConstants.ts]: Ah, ok, got it. A comment briefly explaining this would be helful in the code. Something like:
```js
// Don't relly on 6 hour long JS intervals which can be flaky in various runtime environments.
```

File: packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx]: Coo. Resolving.

File: ci/packages/suite-web.yml
Comment (goodhoko) [reply ci/packages/suite-web.yml]: Thx! Resolving.

File: docs/misc/message_system.md
Comment (goodhoko) [reply docs/misc/message_system.md:58]: It's generally not the best practice to track generated files. Though sometimes simply commiting the file is simpler than ensuring it's automaticaly generated whenever needed.

In our case there's the `build:libs` job in the root package.json of the monorepo that is already required for suite to  build successfuly. (Right?) If we can ensure that `messageSystem.ts` is generated with this script too I'd suggest not to track it in git.

Comment (goodhoko) [reply docs/misc/message_system.md:58]: It's generally not the best practice to track generated files. Though sometimes simply commiting the file is simpler than ensuring it's automaticaly generated whenever needed.

In our case, as you write, there's the `build:libs` job in the root package.json of the monorepo that is already required for suite to  build successfuly. And since `messageSystem.ts` is generated by that job too there's IMO no need to track it in git.

File: packages/suite-data/src/message-system/config/config.v1.json
Comment (tomasklim) [reply packages/suite-data/src/message-system/config/config.v1.json:6]: Yes, definitely. I have a note in the first post.

Btw if you already have something what we want to put into the config, let's do it.

File: docs/misc/message_system.md
Comment (tomasklim) [reply docs/misc/message_system.md]: changed

File: packages/suite/src/actions/suite/messageSystemActions.ts
Comment (slowbackspace) [reply packages/suite/src/actions/suite/messageSystemActions.ts:108]: console.error will be logged in sentry. I'm just afraid nobody will notice anyway, but still could be useful to determine how many users are affected 😁

Comment (goodhoko) [reply packages/suite/src/actions/suite/messageSystemActions.ts:108]: > console.error will be logged in sentry.

Perfect! Good to know.

> I'm just afraid nobody will notices anyway,

Yeah. Reporting to sentry is one thing. Getting these errors in front of the right eyes is another story. .D Where do sentry errors currenrly end up? At GoOut we had a dedicated slack channels where sentry (and other runtime) errors were posted by a bot. Shall I create an issue for such an integration?

I'll resolve this thread as the original concren is addressed.

Comment (slowbackspace) [reply packages/suite/src/actions/suite/messageSystemActions.ts:108]: Reports are only in sentry right now. Some kind of bot is interesting idea, but since we have literally thousands of reports so it could get ugly 😁 Maybe it could report only after certain amount of people experienced the issue or something like that, that would be really cool!

Imo first we need address these reports. So time ago I created github issues from few most common sentry reports (https://github.com/trezor/trezor-suite/issues?q=is%3Aopen+is%3Aissue+label%3ASentry). Another problem is that too many people are using old version  of the app (because of non functioning auto updater) and it is kinda hard to organise all these issues (I blame sentry UI, but it could be that we are just doing something wrong). Anyway It will take quite a lot of time and effort to clean up sentry to it can really start useful and sadly priorities are elsewhere a the moment 😞

Comment (goodhoko) [reply packages/suite/src/actions/suite/messageSystemActions.ts:108]: Aaaah, so we need to get out of the dirty waters first to actually see through to valuable information.

> Maybe it could report only after certain amount of people experienced the issue or something like that, that would be really cool!

At first sight probably yes. At second I'd call that a band aid at best. We're not a service distributed across datacenters all over the globe to ignore erros not reaching an occurence threshold. (OFC occurence should still be used for triage.) I think that we need to aim for 0 errors reported. The closer we get to that the more valuable the reported errors will be. That goal is impossibl of course. But that doesn't stop us from aiming for it. ,)

> Imo first we need address these reports. So time ago I created github issues from few most common sentry reports

This is IMO a perfect start. 👍 I think it wouldn't be a bad idea to include these issues in the Acceptence criteria of Suite official start.

Btw: how do I access the Sentry reports?

File: packages/suite-data/src/message-system/config/config.v1.json
Comment (tsusanka) [reply packages/suite-data/src/message-system/config/config.v1.json:6]: @prusnak anything in our minds that we would like to push via the messaging system already? I have one in my mind:
- Bridge versions [28, 29, 30] on web & Windows.

Comment (tsusanka) [reply packages/suite-data/src/message-system/config/config.v1.json:6]: @tomasklim let's not block this PR because of this. We'll figure it out later!

Comment (tomasklim) [reply packages/suite-data/src/message-system/config/config.v1.json:6]: sure, I am leaving the config there to make it easier for PR reviewers. Before merging, I will replace it with an empty config.

```
{
    "version": 1,
    "timestamp": "2021-04-01T00:00:00+00:00",
    "sequence": 1,
    "actions": []
}
```

File: docs/misc/message_system.md
Comment (tomexx) [inline docs/misc/message_system.md]: probably tabs instead of spaces used

File: docs/misc/message_system.md
Comment (tomexx) [inline docs/misc/message_system.md:195]: It is text-only by design?

File: docs/misc/message_system.md
Comment (tomexx) [inline docs/misc/message_system.md]: I think OR (union) might be more useful. Let's say I want to target Firefox 80 on Mac and Firefox 82 on Windows with a same message. If I understand it correctly, I'd have to create two separate actions if AND is used here.

File: packages/suite-data/src/message-system/config/config.v1.json
Comment (tomexx) [inline packages/suite-data/src/message-system/config/config.v1.json:1]: Is this legit config, or is it just for testing purposes?

File: docs/misc/message_system.md
Comment (tomexx) [inline docs/misc/message_system.md:173]: Is this a random UUID or should it match something?

File: docs/misc/message_system.md
Comment (tomexx) [inline docs/misc/message_system.md]: Let's maybe rename this to `link`

**Review by tomexx — COMMENTED** [review]
Good job + I've got a few questions and meanwhile I'll go now try it locally

File: packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx
Comment (keraf) [inline packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx]: Maybe worth memoizing here? Same for `getDismissalConfig`.

File: packages/suite/src/components/suite/Banners/index.tsx
Comment (keraf) [inline packages/suite/src/components/suite/Banners/index.tsx]: Seems to have quite some shuffling around happening here, maybe memoizing that one too might be a good shout.

**Review by keraf — APPROVED** [review]
That's some really solid work, great job! I just have a few concerns about certain operations that look quite expensive in the rendering of components, I might have missed some (would maybe be worth going over it again).

File: docs/misc/message_system.md
Comment (tomasklim) [reply docs/misc/message_system.md:195]: Yes, it is text-only. I do not think HTML makes sense here.

@tsusanka or does it?

File: docs/misc/message_system.md
Comment (tomasklim) [reply docs/misc/message_system.md:173]: Yes, random UUID. It has to be unique to avoid displaying dismissed messages twice.

File: packages/suite-data/src/message-system/config/config.v1.json
Comment (tomasklim) [reply packages/suite-data/src/message-system/config/config.v1.json:1]: only for testing purposes, will be deleted before merging

File: docs/misc/message_system.md
Comment (tomasklim) [reply docs/misc/message_system.md]: Oh, a bug in the docs. The operation between object conditions is OR indeed.

File: packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx
Comment (tomasklim) [reply packages/suite/src/components/suite/Banners/MessageSystemBanner.tsx]: All three functions memoized. I checked the code and did not find anything else to memoize 👍

File: docs/misc/message_system.md
Comment (tsusanka) [reply docs/misc/message_system.md:195]: Hmm :thinking:. HTML no but I am wondering if markdown wouldn't be nice. But that's out of scope for now, let's ship it and iterate later :).

File: docs/misc/message_system.md
Comment (tomasklim) [reply docs/misc/message_system.md:58]: Generated on ```build:libs``` now. Not commited to repo

File: packages/suite/src/actions/suite/constants/messageSystemConstants.ts
Comment (tomasklim) [reply packages/suite/src/actions/suite/constants/messageSystemConstants.ts]: Added

**Comment by sorooris** [comment]
QA OK

Tested in the May release, details here https://github.com/trezor/trezor-suite/issues/2752#issuecomment-841173784; moving to Approved.

**Pull request merged**
